CodeSnippetsReflection.OpenAPI/LanguageGenerators/GoGenerator.cs

```diff
@@ -15,264 +16,321 @@ public class GoGenerator : ILanguageGenerator<SnippetModel, OpenApiUrlTreeNode>
-                    WriteProperty(payloadSB, property.Value, null, indentManager, propertyAssignment, null, ",");
-                }
-                indentManager.Unindent();
-                payloadSB.AppendLine($"{indentManager.GetIndent()}}}");
+            else
+            {
+                builder.AppendLine($"{indentManager.GetIndent()}{requestBodyVarName} := graphmodels.New{codeGraph.Body.Name}()");
+                WriteCodePropertyObject(requestBodyVarName, builder, codeGraph.Body, indentManager);
             }
         }
-        private static void WritePropertyFromTempVar(StringBuilder payloadSB, string propertyAssignment, string tempVarName, string valueAssignment, string valueDeclaration, bool addPointer, string propertySuffix) {
-            if(!string.IsNullOrEmpty(tempVarName)) {
-                payloadSB.AppendLine($"{tempVarName}{valueAssignment}{valueDeclaration}");
-                var pointer = addPointer ? "&" : string.Empty;
-                payloadSB.AppendLine($"{propertyAssignment}{pointer}{tempVarName}{propertySuffix}");
-            } else {
-                payloadSB.AppendLine($"{propertyAssignment}{valueDeclaration}{propertySuffix}");
+        private static string GetActionParametersList(params string[] parameters)
+        {
+            var nonEmptyParameters = parameters.Where(p => !string.IsNullOrEmpty(p));
+            if (nonEmptyParameters.Any())
+                return string.Join(", ", nonEmptyParameters.Aggregate((a, b) => $"{a}, {b}"));
+            else return string.Empty;
+        }
+
+        private static void WriteArrayProperty(string propertyAssignment, string objectName, StringBuilder builder, CodeProperty parentProperty, CodeProperty codeProperty, IndentManager indentManager){
+            var contentBuilder = new StringBuilder();
+            var propertyName = NormalizeJsonName(codeProperty.Name.ToFirstCharacterLowerCase());
+
+            var objectBuilder = new StringBuilder();
+            IndentManager indentManagerObjects = new IndentManager();
+
+            indentManager.Indent();
+            var childPosition = 0;
+            foreach (var child in codeProperty.Children){
+                if(child.PropertyType == PropertyType.Object){
+                    WriteCodeProperty(propertyAssignment, objectBuilder, codeProperty, child , indentManagerObjects, childPosition);
+                    contentBuilder.AppendLine($"{indentManager.GetIndent()}{child.Name.ToFirstCharacterLowerCase()?.IndexSuffix(childPosition)},");
+                }else{
+                    WriteCodeProperty(propertyAssignment, contentBuilder, codeProperty, child , indentManager, childPosition);
+                }
+                childPosition++;
             }
+            indentManager.Unindent();
+
+            if(objectBuilder.Length > 0){
+                builder.AppendLine("\n");
+                builder.AppendLine(objectBuilder.ToString());
```

Comment:
Avoid calling `ToString()` until it is absolutely needed. There might be optimizations that you can benefit from if you call the variation that takes the StringBuilder directly.
```suggestion
                builder.Append(objectBuilder);
                builder.AppendLine();
```

Same comment goes for all the other instances where the string is passed when passing StringBuilder is possible.

Reply:
This is the only instance for the `objectBuilder` , which is a separate builder from `builder` calls the `ToString`.  the reason was its the `builder` object traverses the object model from the `root` of the tree, but when writing an array, all the instances of the objects to be part of the array need to be generated first then added to the array have a look at this example https://github.com/microsoftgraph/microsoft-graph-devx-api/pull/1021#issuecomment-1151093487

For this reason, a second builder to traverse all the children was required

Reply:
I understand the need for the second builder, my suggestion is not to call `ToString()` on it. The `Append` call that takes `StringBuilder` may handle buffer transfer more efficiently than converting it into `string` and then passing into the outer builder as a string.

Reply:
Ok, I ran the benchmarks. With `ToString` it is faster even though memory allocations are bigger. I don't think we will be memory bound here, so we can keep the code as is:

```
// * Summary *

BenchmarkDotNet=v0.13.1, OS=Windows 10.0.22000
Intel Xeon W-2123 CPU 3.60GHz, 1 CPU, 8 logical and 4 physical cores
.NET SDK=6.0.300
  [Host]     : .NET 6.0.5 (6.0.522.21309), X64 RyuJIT
  DefaultJob : .NET 6.0.5 (6.0.522.21309), X64 RyuJIT


|          Method |     Mean |    Error |   StdDev | Rank |      Gen 0 |     Gen 1 |     Gen 2 | Allocated |
|---------------- |---------:|---------:|---------:|-----:|-----------:|----------:|----------:|----------:|
|    WithToString | 87.59 ms | 1.721 ms | 4.883 ms |    1 | 18857.1429 | 4714.2857 | 1714.2857 |    142 MB |
| WithoutToString | 94.19 ms | 2.830 ms | 8.256 ms |    2 | 19166.6667 | 5000.0000 | 1500.0000 |    131 MB |
```

Benchmark code using BenchmarkDotNet package:
```csharp
using System.Text;
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;

public static class Program
{
    public static void Main(string[] args)
    {
        var summary = BenchmarkRunner.Run<Benchmarker>();
    }

    [MemoryDiagnoser]
    [RankColumn]
    public class Benchmarker
    {
        static int items = 1000000;
        [Benchmark]
        public string WithToString()
        {
            var sb = new StringBuilder();
            for (int i = 0; i < items; i++)
            {
                sb.Append(i.ToString());
            }
            var innerSB = new StringBuilder();
            for (int i = 0; i < items; i++)
            {
                innerSB.Append(i.ToString());
            }
            sb.Append(innerSB.ToString());
            return sb.ToString();
        }

        [Benchmark]
        public string WithoutToString()
        {
            var sb = new StringBuilder();
            for (int i = 0; i < items; i++)
            {
                sb.Append(i.ToString());
            }
            var innerSB = new StringBuilder();
            for (int i = 0; i < items; i++)
            {
                innerSB.Append(i.ToString());
            }
            sb.Append(innerSB);
            return sb.ToString();
        }
    }
}
```

Reply:
Thanks for the detailed explanation